### ISSUE TITLE
typed_per_filter_config for aws lambda is not recognized

### ISSUE THREAD

**Issue by reskin89** [envoyproxy/envoy#26855]
*Title*: *typed_per_filter_config for aws lambda is not recognized*

*Description*:
I have an envoy configuration, that is using `typed_per_filter_config` per route, so I can invoke a different aws lambda per route.  I attempted to use a cluster per route, but `typed_per_filter_config` is an invalid argument according to envoy on a cluster assignment, (I copy and pasted the doc examples).

When using it per route, I get:

```
INVALID_ARGUMENT:(route_config.virtual_hosts[0].routes[1].typed_per_filter_config[0].value): invalid value "type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig" for type type.googleapis.com/google.protobuf.Any
```

[optional *Relevant Links*:]
Here is the configuration:

```yaml
---
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 15000 }

static_resources:
  listeners:
  - name: lambda_listener
    address: 
      socket_address: { address: 127.0.0.1, port_value: 21000 }
    filter_chains:
    - filters:
      -  name: envoy.filters.network.http_connection_manager
         typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
            stat_prefix: ingress_http
            codec_type: AUTO
            route_config:
              name: lambda1
              virtual_hosts:
              - name: local_lambda1
                routes:
                  - match: { prefix: "/lambda1 "}
                    route: { cluster: lambda1_cluster }
                    typed_per_filter_config:  
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                        arn: "arn:aws:lambda:us-east-1:987654321:function:lambda1"
                        payload_passthrough: true
                  - match: { prefix: "/lambda2" }
                    route: { cluster: lambda2_cluster }
                    typed_per_filter_config:  
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                          arn: "arn:aws:lambda:us-east-1:987654321:function:lambda2"
                          payload_passthrough: true
  # clusters:
  # - name: lambda1_cluster
  #   typed_per_filter_config:
  #     envoy.filters.http.aws_lambda:
  #     "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
  #     invoke_config:
  #       arn: "arn:aws:lambda:us-east-1:987654321:function:lambda1"
  #       payload_passthrough: true

  # - name: lambda2_cluster 
  #   typed_per_filter_config:
  #     envoy.filters.http.aws_lambda:
  #     "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
  #     invoke_config:
  #       arn: "arn:aws:lambda:us-east-1:987654321:function:lambda2"
  #       payload_passthrough: true
```

**Comment by suniltheta** [comment]
would you need to specify `envoy.filters.http.aws_lambda:` between `typed_per_filter_config:` &   `"@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig` ?

**Comment by reskin89** [comment]
So that got me past that, but now its telling me since I'm using `match` and action is required, my action is to invoke the lambda, but it doesn't seem to think so, it wants one of: route, redirect, or direct_response.  I'm not sure how to make it work this way but the docs say it can.... so I guess I need to use a cluster? but when I used the clusters it was saying `typed_per_filter_config` is an invalid argment there, which is contrary to the docs...

**Comment by reskin89** [comment]
The docs seem kind of all over, it says [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/aws_lambda/v3/aws_lambda.proto#extensions-filters-http-aws-lambda-v3-perrouteconfig)  that what I'm trying to do should be possible with what I'm configuring, but envoy itself seems to complain.

for reference, I'm now here:

```yaml
---
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 15000 }

static_resources:
  listeners:
  - name: lambda_listener
    address: 
      socket_address: { address: 127.0.0.1, port_value: 21000 }
    filter_chains:
    - filters:
      -  name: envoy.filters.network.http_connection_manager
         typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
            codec_type: AUTO
            stat_prefix: ingress_http
            route_config:
              name: lambda1
              virtual_hosts:
              - name: local_lambda1
                domains: ["*"]
                routes:
                  - match: { path: "/lambda1 "}
                    typed_per_filter_config:  
                      envoy.filters.http.aws_lambda:
                        "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                        invoke_config:
                          arn: "arn:aws:lambda:us-east-1:987654321:function:lambda1"
                          payload_passthrough: true
            http_filters:
            - name: envoy.filters.http.router
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
            - name: envoy.filters.http.aws_lambda
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
```

If I'm reading the docs right, I should be able to do this per route per virtual host, but I get the error:

`Proto constraint validation failed (HttpConnectionManagerValidationError.RouteConfig: embedded message failed validation | caused by RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Routes[0]: embedded message failed validation | caused by field: "action", reason: is required): stat_prefix: "ingress_http"`

The docs also say `stat_prefix` isn't required but it is.  "Action" seems to be referring to `route`, `redirect`, or `direct_response`, it seems that invocation doesn't count as an action.

**Comment by suniltheta** [comment]
In the first config that you share it didn't have `http_filters`. Can you try by including the `envoy.filters.http.router` and not `envoy.filters.http.aws_lambda` under `http_filters`, while still keeping `envoy.filters.http.aws_lambda` under typed_per_filter_config?

**Comment by reskin89** [comment]
I did which still results in:
```
Proto constraint validation failed (HttpConnectionManagerValidationError.RouteConfig: embedded message failed validation | caused by RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Routes[0]: embedded message failed validation | caused by field: "action", reason: is required): stat_prefix: "ingress_http"
```

**Comment by reskin89** [comment]
So I think I found a nesting issue, and fixed the cluster config and now am receiving:

```shell
<MissingAuthenticationTokenException>
  <Message>Missing Authentication Token</Message>
</MissingAuthenticationTokenException>
```

back from AWS.  I can't seem to see in the docs where auth is provided.  I have my env authenticated and can perform aws cli commands in the same session I'm running envoy.  Where would these configurations be provided?

**Comment by suniltheta** [comment]
Can you share the full config so that in future we can refer what was the missing piece? Or you can share what was done wrong.

---

Can you share Envoy debug logs with grep on `aws` & `misc` to see if they point to why the auth is missing.

The auth is picked form Env variables, EC2 task role or ECS task metadata. Where is your application running?

**Comment by reskin89** [comment]
Sure, this is the working configuration, it seems there was an odd extra space in the nesting, and envoy was still loading the config, but not really loading it per-say.  I'm running envoy locally on a macbook for testing, and have my `AWS_PROFILE` and `AWS_REGION` set in my environment.  I'm using [func-e](https://func-e.io/) to run envoy, but I don't think it runs it in any type of "isolation" or container, so my environment should be seen.

```yaml
---
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 15000 }

static_resources:
  listeners:
  - name: lambda_listener
    address: 
      socket_address: { address: 0.0.0.0, port_value: 8080 }
    filter_chains:
    - filters:
      -  name: envoy.filters.network.http_connection_manager
         typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: ingress_http
          route_config:
            name: lambda1
            virtual_hosts:
            - name: local_lambda1
              domains: ["*"]
              routes:
                - match: { prefix: "/lambda1" }
                  route: { cluster: lambda1_cluster }
                  typed_per_filter_config:  
                    envoy.filters.http.aws_lambda:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                        arn: "arn:aws:lambda:us-east-1:987654321:function:lambda1"
                        payload_passthrough: true
                - match: { prefix: "/lambda2" }
                  route: { cluster: lambda1_cluster }
                  typed_per_filter_config:  
                    envoy.filters.http.aws_lambda:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                        arn: "arn:aws:lambda:us-east-1:987654321:function:lambda2"
                        payload_passthrough: true
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router

  clusters:
  - name: lambda1_cluster
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    metadata:
      filter_metadata:
        com.amazonaws.lambda:
          egress_gateway: true
    load_assignment:
      cluster_name: lambda1_cluster # does this have to match? seems redundant
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: lambda.us-east-1.amazonaws.com
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: "*.amazonaws.com"
```

**Comment by suniltheta** [comment]
This file does the job of loading all the credentials https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/aws/credentials_provider_impl.cc#L254-L281

It will try to get credentials from `AWS_ACCESS_KEY_ID` & `AWS_SECRET_ACCESS_KEY` env variables. If those are not set then you would not be able to authenticate from Mac.

There is an improvement coming to load creds form credentials file https://github.com/envoyproxy/envoy/pull/25532 but that will not be available yet on func-e until v1.27 release

**Comment by suniltheta** [comment]
I see the issue with envoy lambda filter documentation  https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/aws_lambda_filter . Please feel free to propose an update.

**Comment by reskin89** [comment]
So I have now added:

`AWS_ACCESS_KEY_ID`
`AWS_SECRET_ACCESS_KEY`
`AWS_REGION`
`AWS_SESSION_TOKEN`

and modified the lambda ARNs to real lambdas and envoy is still not utilizing these credentials.

I could build an ECS task to do this but I figured testing with func-e would be much much faster and easier to iterate on.

**Comment by suniltheta** [comment]
can you confirm if this issue is also seen when you use aws_lambda but don't use `PerRouteConfig` lambda invocation ?

**Comment by reskin89** [comment]
Could you clarify what you mean by that? I.e. by setting the `http_filter` so all routes are affected equally?

**Comment by suniltheta** [comment]
true, I just want to see if the credentials issues is restricted to just PerRouteConfig. I have tested the other scenario and the credentials works for me. I have not tested the PerRouteConfig scenario.

**Comment by reskin89** [comment]
So it looks like that caused it to load the credentials properly, however it seems like its loading something that's stale?

I can run `aws lambda invoke --function-name MYARN` without issue in the same environment, however when envoy runs and attempts to invoke function AWS is returning a "stale credential provided" error.


By this is seems that PerRouteconfig does *not* actually load the aws credentials, which eliminates the ability to have a lambda invoked per route.  The latest version this was tested on so far is 1.25.4.

**Comment by reskin89** [comment]
For full scope sake, here's the current config that I'm receiving stale credentials on when running locally with the keys in my environment:

```yaml
---
admin:
  address:
    socket_address: { address: 127.0.0.1, port_value: 15000 }

static_resources:
  listeners:
  - name: lambda_listener
    address: 
      socket_address: { address: 0.0.0.0, port_value: 8080 }
    filter_chains:
    - filters:
      -  name: envoy.filters.network.http_connection_manager
         typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: ingress_http
          route_config:
            name: lambda1
            virtual_hosts:
            - name: local_lambda1
              domains: ["*"]
              routes:
                - match: { path: "/health" }
                  direct_response:
                    status: 200
                    body:
                      inline_string: '{ "message":"healthy" }'
                - match: { prefix: "/lambda1" }
                  route: { cluster: lambda1_cluster }
          http_filters:
          - name: envoy.filters.http.aws_lambda
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.Config
              arn: "arn:aws:lambda:us-east-1:REDACTED:function:mesh-direct-invoke-lambda1"
              payload_passthrough: false
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          
  clusters:
  - name: lambda1_cluster
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    metadata:
      filter_metadata:
        com.amazonaws.lambda:
          egress_gateway: true
    load_assignment:
      cluster_name: lambda1_cluster # does this have to match? seems redundant
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: lambda.us-east-1.amazonaws.com
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: "*.amazonaws.com"
```

**Comment by suniltheta** [comment]
I can reproduce this bug on my end if I invoke PerRouteConfig. For regular invocation it is able to pick the credentials.
```
< 
<MissingAuthenticationTokenException>
  <Message>Missing Authentication Token</Message>
</MissingAuthenticationTokenException>
* Connection #0 to host localhost left intact
```

**Comment by reskin89** [comment]
I'm not really too versed in C, is this a simple bug fix by chance?  The feature is becoming a bit critical to me.  I'd dive in and PR myself if I understood a bit of the code.

**Comment by suniltheta** [comment]
I need to dig into code where it is failing to fetch credentials for per route config. I would look into that hopefully by later today.

meanwhile from the latest config that you shared, I have almost same config except I had to change host_rewrite_literal like below. See if you include this host_rewrite_literal it works (Not the PerRouteConfig but for the all routes in the filter chain)

```
              routes:
                - match:
                    prefix: "/"
                  route:
                    cluster: lambda_cluster
                    host_rewrite_literal: lambda.us-east-1.amazonaws.com
```

**Comment by reskin89** [comment]
so, even with host_rewrite_literal, I'm getting:
```
{"message":"The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details."}
```

and I know the keys I have in my environment, and in ECS are valid.  ECS has a role that gives it more permissions than it should admittedly, including `lambda:*`

Thank you for looking at this!

**Comment by suniltheta** [comment]
See if there are any unintended headers that are included in the request that is sigv4 signed. You should be able to see it from envoy debug level logs

**Comment by reskin89** [comment]
There aren't any requests that are sigv4 signed, however it does look like its trying to load AWS_SESSION_TOKEN even if its unset and unneeded:

```
[2023-04-21 13:30:14.187][1285700][debug][aws] [source/extensions/common/aws/credentials_provider_impl.cc:60] Found following AWS credentials in the environment: AWS_ACCESS_KEY_ID=REDACTED, AWS_SECRET_ACCESS_KEY=*****, AWS_SESSION_TOKEN=
```

**Comment by suniltheta** [comment]
Does the canonical request that is signed include a header with empty SecurityToken?

https://github.com/envoyproxy/envoy/blob/c115053144eb29f4c45093ef482c236b211142eb/source/extensions/common/aws/signer_impl.cc#L57-L59

**Comment by reskin89** [comment]
Here is the request:

```shell
[2023-04-21 13:44:27.724][1295807][debug][aws] [source/extensions/common/aws/signer_impl.cc:68] Canonical request:
POST
/2015-03-31/functions/arn%3Aaws%3Alambda%3Aus-east-1%3AREDACTED%3Afunction%3Amesh-direct-invoke-lambda1/invocations

accept:*/*
content-length:238
content-type:application/json
host:127.0.0.1:8080
user-agent:curl/7.77.0
x-amz-content-sha256:9928ef2def0d92f7ecd6fae542f8bfc94fb237c7edb47c3ef975e67cc461e011
x-amz-date:20230421T174400Z
x-amz-invocation-type:RequestResponse
x-request-id:987d0bc4-2fea-4b2c-85da-4764a2671f30

accept;content-length;content-type;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-invocation-type;x-request-id
9928ef2def0d92f7ecd6fae542f8bfc94fb237c7edb47c3ef975e67cc461e011
```

**Comment by suniltheta** [comment]
Can you try by specifying the host header when you make a call?

```
curl -H "Host: lambda.us-east-1.amazonaws.com" localhost:8080
```

---

I believe this should have worked without specifying the lambda host header. I don't know if it is a bug where it signs the request with host header that is not intended.

**Comment by reskin89** [comment]
Adding the host header resolved the signing issue entirely! I didn't even notice that it was signing with a host header of a localhost/loopback.  That's probably not going to be received well from amazon.

**Comment by suniltheta** [comment]
I see that to make `aws_lambda.v3.PerRouteConfig` to work without getting `MissingAuthenticationTokenException` you would have to include the `aws_lambda.v3.Config` under `http_filters` as well.

**Comment by reskin89** [comment]
Would you have an example? Isn't that configuring the global route filter for one lambda for all routes to the given cluster?

**Comment by suniltheta** [comment]
but I believe PerRouteConfig should override the global route filter. There was a bug https://github.com/envoyproxy/envoy/issues/23076 fix made recently related to this where signing filter was made to pick the region from the arn of PerRouteConfig.

**Comment by reskin89** [comment]
Awesome, that does work!

Now my very last piece.  `host_rewrite_literal` doesn't actually seem to be rewriting the host header.  Is there another way to do this? Otherwise clients that utilize this for me will have to pass a host header into every request

**Comment by reskin89** [comment]
I've also attempted to use host_rewrite on the route and that doesn't seem to change the host header either

**Comment by reskin89** [comment]
It looks like `auto_host_rewrite` may not be compatible with a `logical_dns` type cluster:

https://github.com/envoyproxy/envoy/issues/13312#issuecomment-704720191

I saw a PR from 2017 that added `auto_host_rewrite` and `host_rewrite_literal` to logical and strict dns cluster types but it doesn't seem to be working. I did add the host rewrite literal to one of my routes as a typed per filter config to no avail, as well as setting it on the filters at the global virtual host layer as well.

Part of me is thinking the lambda filter may just be taking priority, but I made it the least priority filter in the route config to no avail.

As one last test, I made the cluster static and just picked one of the many IP addresses for lambda.us-east-1.amazonaws.com and set it as the address for the endpoint.

It seems `auto_host_rewrite`, and `host_rewrite_literal` both still fail there, which seems odd and makes me think I'm missing something.

Here's the updated config:

```yaml
static_resources:
  listeners:
  - name: lambda_listener
    address: 
      socket_address: { address: 0.0.0.0, port_value: 8080 }
    filter_chains:
    - filters:
      -  name: envoy.filters.network.http_connection_manager
         typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: ingress_http
          route_config:
            name: lambda1
            virtual_hosts:
            - name: local_lambda1
              domains: ["*"]
              routes:
                - match: { prefix: "/lambda1" }
                  route:
                    cluster: lambda1_cluster
                    host_rewrite_literal: lambda.us-east-1.amazonaws.com
                  typed_per_filter_config:
                    envoy.filters.http.aws_lambda:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                        arn: "arn:aws:lambda:us-east-1:REDACTED:function:mesh-direct-invoke-lambda1"
                        payload_passthrough: true
                - match: { prefix: "/lambda2" }
                  route: 
                    cluster: lambda1_cluster 
                  typed_per_filter_config:  
                    envoy.filters.http.aws_lambda:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.PerRouteConfig
                      invoke_config:
                        arn: "arn:aws:lambda:us-east-1:REDACTED:function:mesh-direct-invoke-lambda2"
                        payload_passthrough: true
          http_filters:
          - name: envoy.filters.http.aws_lambda
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.aws_lambda.v3.Config
              arn: "arn:aws:lambda:us-east-1:395284687305:function:mesh-direct-invoke-lambda1"
              payload_passthrough: true
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router

  clusters:
  - name: lambda1_cluster
    connect_timeout: 0.25s
    type: STATIC
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    metadata:
      filter_metadata:
        com.amazonaws.lambda:
          egress_gateway: true
    load_assignment:
      cluster_name: lambda1_cluster # does this have to match? seems redundant
      endpoints:
      - lb_endpoints:
        - endpoint:
            hostname: lambda.us-east-1.amazonaws.com
            address:
              socket_address:
                address: 100.25.113.238
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        sni: "*.amazonaws.com"
```

**Comment by reskin89** [comment]
I am indeed seeing:

```
[2023-04-24 14:56:45.654][3393667][debug][aws] [source/extensions/common/aws/signer_impl.cc:79] Signing request with: AWS4-HMAC-SHA256 Credential=REDACTED/20230424/us-east-1/lambda/aws4_request, SignedHeaders=accept;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-invocation-type;x-request-id, Signature=REDACTED
```

Before any other filters are applied, directly after the canonical request output.  I've attached the log for your review as well.
[envoylambdalog.log](https://github.com/envoyproxy/envoy/files/11313746/envoylambdalog.log)

**Comment by suniltheta** [comment]
I guess it doesn't make sense to include the incoming host header `host:127.0.0.1:8080` for signing request to aws lambda. We need to figure out if this is a bug that needs to be fixed or this was intentionally the original design.

**Comment by reskin89** [comment]
Yeah, I'm also wondering why it seemingly can't be overridden at all.  I don't see how this integration can work without someone providing the host header in their call

**Comment by suniltheta** [comment]
I see that the problem is with including the port number in the host header.

```curl -H "Host: localhost" localhost:8080

**Comment by suniltheta** [comment]
I guess the fix can be to sanitize the host header while including it for signing the request

**Comment by reskin89** [comment]
That seems like the simplest route to go. Id still wonder why other filters can't do anything but that's more of a curiosity and doesn't seem crucial to functionality

**Comment by reskin89** [comment]
With my limited C++ knowledge I tried to achieve this and am honestly not sure how to access the host header in the lambda signing functions. I know it's there in the http message it's handed 🤦

**Comment by reskin89** [comment]
So, I'm not so sure its the port.  As a test I set my listener config to port 80 so I could make my call without the port:

`host:127.0.0.1`
`host:localhost`

and I'm still left with an invalid signature.  I even appended the host header on my curl `"Host: localhost"` and was met with the same until I applied the Host Header as `lambda.REGION.amazonaws.com`

**Comment by reskin89** [comment]
@suniltheta could you point me in the lambda code to where the headers are set?

I'm going to try to create a fix where it sets the host header to the lambda endpoint configured to try and solve it and submit a PR.

Thanks again!

**Comment by suniltheta** [comment]
We need to make sure that the header before signing is sanitized and doesn't contain the port number.
The signing `sigv4_signer_->sign*` is called inside 2 functions `decodeHeaders` & `decodeData`

I would need to check by adding some debug logs to check why host_rewrite_literal is not taking into effect before lambda filter is applied.

**Comment by suniltheta** [comment]
On checking https://www.envoyproxy.io/docs/envoy/v1.26.1/api-v3/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto it has host_rewrite config option. Maybe something like that can be introduced?

**Comment by reskin89** [comment]
I had set my listener to port 80 so I didn't have to include the port number in my curl and it resulted in the same error, I'm not so sure it's the port number as much as the target host that the lambda endpoint receives is "localhost" and not "lambda.region.amazonaws.com"

**Comment by reskin89** [comment]
@suniltheta I have submitted PR #27551 to attempt to fix this issue.  Instead of removing the port, it seemed more prudent to rewrite the host header to `lambda.REGION.amazonaws.com` as that's what the host header should be when making a call to the lambda api endpoint.

**Comment by juanmolle** [comment]
@suniltheta I fall in the same issue with auto_host_rewrite. the signature is using localhost instead the overrider host.
Is there any update on this issue? Should I create a new one?

**Comment by reskin89** [comment]
There is a PR that I need to amend and resubmit to fix this, I'm going to try and get back to that this week

**Comment by juanmolle** [comment]
@reskin89 I see a message to call for low risk fixes yesterday due to 1.29 is near to release. Let me know if you need a double check of the fix to try to target this next release.
